apps/test-app/frontend/src/components/tree-widget/TreeWidget.tsx

```diff
@@ -139,43 +128,136 @@ export function RulesDrivenTreeWidget(props: Props) {
+  );
+
+  useEffect(() => {
+    if (metadataProvider && queryExecutor) {
+      const sharedProps = {
+        metadataProvider,
+        hierarchyDefinition: new ModelsTreeDefinition({ metadataProvider }),
+        queryExecutor,
+      };
+      if (value) {
+        setFilteringStatus(FilteringInputStatus.FilteringFinished);
+        setModelsTreeHierarchyProvider(
+          new HierarchyProvider({
+            ...sharedProps,
+            filtering: {
+              paths: value,
+            },
+          }),
+        );
+      } else {
+        setModelsTreeHierarchyProvider(new HierarchyProvider(sharedProps));
+      }
+    }
+  }, [queryExecutor, metadataProvider, value]);
+
+  const dataProvider = useMemo((): TreeDataProvider => {
     return async (node?: TreeNodeItem): Promise<TreeNodeItem[]> => {
       const parent: HierarchyNode | undefined = node ? (node as any).__internal : undefined;
       try {
-        return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+        if (modelsTreeHierarchyProvider) {
+          return (await modelsTreeHierarchyProvider.getNodes({ parentNode: parent })).map(parseTreeNodeItem);
+        }
+        return [];
       } catch (e) {
         // eslint-disable-next-line no-console
         console.error(e);
         return [];
       }
     };
-  }, [props.imodel]);
+  }, [modelsTreeHierarchyProvider]);
   const modelSource = useTreeModelSource(dataProvider);
   const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
   const eventHandler = useMemo(() => new TreeEventHandler({ nodeLoader, modelSource }), [nodeLoader, modelSource]);
   const treeModel = useTreeModel(modelSource);
+
+  const [heightToUse, setHeightToUse] = useState(0);
+  const treeWidgetHeaderRef = useRef<HTMLDivElement>(null);
+  useEffect(() => {
+    const heightOfHeader = treeWidgetHeaderRef.current?.clientHeight ?? 0;
+    const heightToSet = props.height ? props.height - heightOfHeader : 0;
+    setHeightToUse(heightToSet ?? 0);
+  }, [props.height]);
+  return (
+    <>
+      <TreeWidgetHeader setFilter={setFilter} filteringStatus={filteringStatus} showFilteringInput={true} ref={treeWidgetHeaderRef} />
+      <div className="filtered-tree">
+        {heightToUse && props.width && (
+          <ControlledTree
+            model={treeModel}
+            eventsHandler={eventHandler}
+            nodeLoader={nodeLoader}
+            selectionMode={SelectionMode.Extended}
+            iconsEnabled={true}
+            width={props.width}
+            height={heightToUse}
+          />
+        )}
+        {filteringStatus === FilteringInputStatus.FilteringInProgress ? <div className="filtered-tree-overlay" /> : null}
+      </div>
+    </>
+  );
+}
+
+interface HeaderProps {
+  setFilter: React.Dispatch<React.SetStateAction<string>>;
+  filteringStatus: FilteringInputStatus;
+  showFilteringInput: boolean;
+  ref: React.RefObject<HTMLDivElement>;
```

Comment:
Use [`forwarfRef`](https://react.dev/reference/react/forwardRef) for exposing ref to parent component

Reply:
Changed